### ISSUE TITLE
Migrate AutoAugment and RandAugment to TensorFlow Addons.

### ISSUE THREAD

**Issue by dynamicwebpaige** [tensorflow/addons#1226]
**Describe the feature and the current behavior/state.**
RandAugment and AutoAugment are both policies for enhanced image preprocessing that are included in EfficientNet, but are still using `tf.contrib`.

https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py

The only `tf.contrib` image operations that they use, however, are [rotate](https://www.tensorflow.org/addons/api_docs/python/tfa/image/rotate), [translate](https://www.tensorflow.org/addons/api_docs/python/tfa/image/translate) and [transform](https://www.tensorflow.org/addons/api_docs/python/tfa/image/transform) - all of which have been included in TensorFlow Addons.

**Relevant information**
- Are you willing to contribute it (yes/no): 
No, but am hoping that someone from the community will pick it up (potentially a Google Summer of Code student)?

- Are you willing to maintain it going forward? (yes/no):
Yes

- Is there a relevant academic paper? (if so, where):
AutoAugment Reference: https://arxiv.org/abs/1805.09501
RandAugment Reference: https://arxiv.org/abs/1909.13719

- Is there already an implementation in another framework? (if so, where):
See link above; this would be a standard migration from `tf.contrib`.

- Was it part of tf.contrib? (if so, where):
Yes

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Image

**Who will benefit with this feature?**
Anyone doing image preprocessing, especially for EfficientNet.

**Comment by bhack** [comment]
For the Gsoc probably we need to think more generally about preprocessing.
We have still numpy based operations in https://keras.io/preprocessing/image/.
Also It needs to be investigated if there will be any integration between AutoAugment/Autokeras.

**Comment by steph-en-m** [comment]
> For the Gsoc probably we need to think more generally about preprocessing.
> We have still numpy based operations in https://keras.io/preprocessing/image/.

@bhack can you please elaborate more on what you mean by this. Thanks

**Comment by bhack** [comment]
Many images operations in Keras are still not in addons or in `tf.image` but numpy implemented:
https://github.com/keras-team/keras-preprocessing/tree/master/keras_preprocessing/image
Other image operations are going in [TF.IO](https://github.com/tensorflow/io/issues/825) or TF.graphics?
I think that we need to unify image processing so that eventually ops could benefit from the compiler stack soon or later (MLIR & friends).

**Comment by bhack** [comment]
Also about AutoKeras can the policy be handled by Autokeras so that autoaugment could be used more generally in other projects/experiment with Autokeras+keras_preprocessing instead of being embedded in Efficientnet?

**Comment by abhichou4** [comment]
> Also about AutoKeras can the policy be handled by Autokeras so that autoaugment could be used more generally in other projects/experiments with Autokeras+keras_preprocessing instead of being embedded in Efficientnet?

I don't think `AutoKeras` handles policies as such. And if policies are just a series of operations, we could store them, maybe as a named tuple. A parser (for the lack of a better word) could then return a `ImageAugmentation` object using a policy. The methods of this class could use Keras preprocessing, `tf.image` etc. Seems messy. Is this better than porting already implemented image ops in tfa.image? 

We could train models with various policies like this as proof of concept.

**Comment by bhack** [comment]
`AutoKeras` was "recently" already integrated with keras preprocessing https://github.com/keras-team/autokeras/pull/922/files.
I agree that probably we can investigate what is needed to expand over Autoaugment (or other kind of) policies. 
But I think that in AutoAugmentation it is also important to have ops that could benefit maximum performance (also in scheduling) cause you will not do these operation offline.

**Comment by fsx950223** [comment]
I'm interested in it and I have benefited a lot from ```AutoAugment```.

**Comment by seanpmorgan** [comment]
So while looking at how AutoKeras is handling hyperparameter containers, at first glance it seems like a suitable replacement for Hparams:
https://github.com/keras-team/keras-tuner/blob/master/kerastuner/engine/hyperparameters.py#L464

We initially decided not to move HParams when we left tf.contrib (even though its convient and works very well) since we didn't want to diverge from officially support APIs.

Adding KerasTuner as a dependency has its own challenges, but I'm wondering if there is a better way to align with the eco-system though using it.

@omalleyt12 @gabrieldemarmiesse Do you have any thoughts about use re-using KerasTuner's HyperParameter object. Or just general thoughts on how AutoAugment in addons fits with the KerasTuner and Keras Preprocessing advances?

**Comment by bhack** [comment]
We still have also AutoAugmentation variant policy in https://github.com/google-research/remixmatch

**Comment by abhichou4** [comment]
Should these augmentations be added as part of a submodule in addons?

**Comment by bhack** [comment]
I don't know. Probably.. It had its own sub "library" for augmentation :smile:  https://github.com/google-research/remixmatch/tree/master/libml
It is quite standard to find fragmentation under google-research repos.

**Comment by carlini** [comment]
The remixmatch repository is intended to faithfully reproduce the the experiments of the corresponding ICLR'20 paper. We do intend for this repository to be the source of truth for any particular implementation.

**Comment by gabrieldemarmiesse** [comment]
@seanpmorgan I don't think that we should depend on Keras-Tuner, except when implementing papers where the end result is a hyperparameter search algorithm. 

For papers that use a search algorithm to produce a result, like AutoAugment and RandAugment, we should hardcode the final numbers [here](https://github.com/tensorflow/tpu/blob/d6f2ef3edfeb4b1c2039b81014dc5271a7753832/models/official/efficientnet/autoaugment.py#L36) and [here](https://github.com/tensorflow/tpu/blob/d6f2ef3edfeb4b1c2039b81014dc5271a7753832/models/official/efficientnet/autoaugment.py#L659) and make sure our API/architecture is modular enough for other people to plug it into an hyperparameter search algorithm.

In short, let's make it easy for users to change those values, it's up to them to plug our API in keras-tuner if they want.

**Comment by bhack** [comment]
@gabrieldemarmiesse What do you think about Remixmatch AutoAugment variant that learns the policy on training? 
https://github.com/google-research/remixmatch/blob/master/libml/augment.py

**Comment by gabrieldemarmiesse** [comment]
@bhack could you expand? I'm not sure I understand your question.

**Comment by bhack** [comment]
@gabrieldemarmiesse I meant how we could generally organize policies as I think that with just a first AutoAugment variant I think we would organize polices a little bit.

**Comment by carlini** [comment]
ReMixMatch's augment policy (CTA) is slightly different from standard augmentation policies because it requires to integrate it with the training loop. At every minibatch step, the policy needs to be "trained" with a second minibatch of examples so that it can determine the magnitude of perturbations that are allowed.

cc @david-berthelot

**Comment by bhack** [comment]
Yes we need to consider also https://github.com/google-research/fixmatch but it seems to me it doesn't introduce new augmentation policies.

**Comment by bhack** [comment]
AutoAugment is also (is it just a copy?) in today released EfficientDet https://github.com/google/automl/blob/master/efficientdet/aug/autoaugment.py. 
We really need a place, like this one, to handle an official AutoAugment extendible API. :smile:

**Comment by carlini** [comment]
Fixmatch uses the same augmentations as ReMixMatch.

**Comment by bhack** [comment]
Thank you for the FixMatch confirmation. Instead I see that @mingxingtan in EfficientDet added some extra AutoAugmentations

**Comment by vinayvr11** [comment]
Hello i want to work on this i am applying for this year GSOC. So did all of the operations are present in Tensorflow Addons. Or we need to create them from scratch.

**Comment by bhack** [comment]
@vinayvr11 There is an inital PR at https://github.com/tensorflow/addons/pull/1275. 
If you can start contribute some ops with PR before be evalauted for GSOC It is better. 
Please open a ticket and mention @abhichou4  to not overlap with him before you start.

**Comment by abhichou4** [comment]
I'll make an issue regarding this, listing all image ops to add. We can also discuss how they can be handled more generally in `tfa.image`.

**Comment by vinayvr11** [comment]
Hello @bhack i have added three more operations to Addons now i want to know that do we have more image operations that are not mention in the list or the list have all the operations.

**Comment by bhack** [comment]
@vinayvr11 Can you post here the ops that are still not in list?

**Comment by vinayvr11** [comment]
@bhack : These are the  operations which are remaining-
to_grayscale
color_jitter
color_jitter_nonrand
color_jitter_rand
compute_crop_shape
center_crop
distorted_bounding_box_crop
crop_and_resize
gaussian_blur
random_crop_with_resize
random_color_jitter
random_blur
batch_random_blur
Shall i also add the bbox operations

**Comment by bhack** [comment]
You can open a new ticket and mention It here after you have checked that osp are not already listed in https://github.com/tensorflow/addons/issues/1333.
I suggest you to collect a bbox operations separate issue and mention It here.
Double check all the source that be have mentioned if there is any other missing ops.

**Comment by vinayvr11** [comment]
Sure

**Comment by vinayvr11** [comment]
> You can open a new ticket and mention It here after you have checked that osp are not already listed in #1333.
> I suggest you to collect a bbox operations separate issue and mention It here.
> Double check all the source that be have mentioned if there is any other missing ops.

Some ops are not checked in https://github.com/tensorflow/addons/issues/1333 so can i add them also in new ticket.

**Comment by abhichou4** [comment]
> > You can open a new ticket and mention It here after you have checked that osp are not already listed in #1333.
> > I suggest you to collect a bbox operations separate issue and mention It here.
> > Double check all the source that be have mentioned if there is any other missing ops.
> 
> Some ops are not checked in #1333 so can i add them also in new ticket.

I'll make sure the rest of the color operations get merged. It would be neat to have another ticket for just bounding box and crop operations.

**Comment by vinayvr11** [comment]
I just created a new ticket for image processing and crop operations.

**Comment by vinayvr11** [comment]
All the tickets are created @bhack.

**Comment by vinayvr11** [comment]
@bhack i want to know something if we added all these images operations now then what we have left for GSOC.

**Comment by bhack** [comment]
I suppose it will be hard to merge all these ops before the application date end but Mentors could use the PR activities to best evaluate candidates.
Also there Is still all the activity about the policy and hyperparam API design for so we could have a pluggable system that It could be valid also for policies that interact with the train loop etc..
If the todo will be not enough for a proposal you could add some interesting image processing operations available in pythorch i.e. [Kornia](https://github.com/kornia/kornia)

**Comment by vinayvr11** [comment]
Ok Thank you i will check it and can i also take some operations from pillow or opencv.

**Comment by bhack** [comment]
@seanpmorgan @gabrieldemarmiesse We need  still if think if colorspace conversions are inside addon or io namaspace. See https://github.com/tensorflow/io/issues/825

**Comment by vinayvr11** [comment]
@bhack : i have created the tickets but i don't know that who will merge them to the repository could you help me this.
Thank you

**Comment by bhack** [comment]
@vinayvr11 Don't worry cause all non Google employees are volunteers here. As soon as there is a free slot we will go ahead with the reviews.

**Comment by vinayvr11** [comment]
Ok thank you @bhack.

**Comment by vinayvr11** [comment]
Hello @bhack could you please help me to find out some more issues in Tensorflow Addons or tf.image() so that i can mention them in GSOC.
Thank you

**Comment by bhack** [comment]
@vinayvr11 these are general hints for every studen.
I suppose that all the image ops and bboxs ops the we have listed from the sparse repositoriees plus the design of an augmentation policy API to porting the policies of the relative repositories that we have mentioned  could be enough for a proposal.
There are also some colorspace conversions that I mentioned if you want to expand the ops coverave and there are also some operations in [Mediapipe](https://github.com/google/mediapipe/tree/master/mediapipe/calculators) calculators that could be interesting to see if could be covered in addon.
For all the applicants I suggested to study a little bit the  policy papers that we have mentioned and the relative repositories to try to make a credibile estimation of the amount of work and figure out a timeline in the proposal.
Having a credibile roadmap in the proposal is a positive evaluation point for the Mentor to figure out that you have really understand the nature of the work that needs to be done in the GSOC.

I suggest you also to partially go ahead, as you can, with PR so that they have a valid sample of your coding. I.e. If possibile take an operator that it Is not already already expressed in Tensorflow (i.e. Numpy/PIL) in these referenced repositories so that Mentors could have a feedback about Tensorflow coding ability other than porting.

**Comment by vinayvr11** [comment]
Thank you very much for this @bhack. Actually i also found some loss functions and optimizers that are listed in tf.contrib but not in Tensorflow 2.x so can i also mention them in proposal.

**Comment by vinayvr11** [comment]
@bhack : could you please review my proposal this will be a great help for me https://docs.google.com/document/d/1mv32xoGI08JP1wcMiugTyVBzCee7dsyYK_5Uf6SjxEE/edit?usp=sharing

**Comment by bhack** [comment]
@vinayvr11 See @dynamicwebpaige [Best Practice](https://docs.google.com/document/d/1zT57PFMGZ04A4CvHxAKVpMTgXjsO92_oKeSKwZMc0Gs/edit) on how to collect a feedback.

**Comment by bhack** [comment]
@dynamicwebpaige I don't know how many slots we could have on similar tasks at GSOC but another related "proxy task" could be image text augmentation like CVPR 2020 https://github.com/Canjie-Luo/Text-Image-Augmentation/

**Comment by gabrieldemarmiesse** [comment]
@bhack @vinayvr11 please keep this thread focused on autoaugment and randaugment. Feel free to use direct messages or to open new issues if you think the topic changed. Having an issue with 40+ messages make the life of the maintainers quite hard.

**Comment by bhack** [comment]
@gabrieldemarmiesse probably was better to open a Gitter channel deicsted to Gsoc other than Gitter addons for this kind of threads to not force ISSUES to go off-topic. Google doesn't official support any realtime chat channel and If you see the Google Summer of code Tensorflow page still points to https://github.com/tensorflow/community (you find also off-topic Gsoc issue there).
Also that repo is mainly used for official RFC also if there Is not a real policy for that repo for ISSUES.
Also IHMO this ISSUE Is quite special, cause involved GSOC in the description and started wtih a very partial overview for a GSOC  proposal related to image transformations and polcies. As we have seen just refencing some other repos quickly emergerd the fragmentation of independet Google teams working on this topic.
I think here we are more interested about a general approach to the transformation and policies and IHMO it Is a better target for a GSOC proposal to have a complete overivew other then porting code. 
So this moved forward a more general discussion. 
For the operations as you seen we have already independent tickets and PR in addons to track the work.

**Comment by tanzhenyu** [comment]
image preprocessing ops only contain basic operations like rotate and translate, it doesn't have all the 'advanced' ops like blend, bbox, or augment.
That said, it might be good to have a standalone repo like keras-image with useful layers / utils that relies on tfa.image?

**Comment by bhack** [comment]
I think we need to de-duplicate code as possible not just cause It Is a waste of time but also cause It Is confusing for newcomers (users and candidate contributors).
I.e. If we collect image ops here in TFA i suppose that the natural upstream Is tf.image as in our migration to core process documentation. 
What about the mentioned policies in this thread? Are these in the addons perimeter? Or in the Keras standalone perimeter? Or in https://github.com/keras-team/autokeras/commit/49281da0a11cd7137818b30a941e88759b6e4741?

**Comment by tanzhenyu** [comment]
> I think we need to de-duplicate code as possible not just cause It Is a waste of time but also cause It Is confusing for newcomers (users and candidate contributors).
> I.e. If we collect image ops here in TFA i suppose that the natural upstream Is tf.image as in our migration to core process documentation.
Why upstream? Sounds like tfa.image is a complimentary suite of tf.image?
> What about the mentioned policies in this thread? Are these in the addons perimeter? Or in the Keras standalone perimeter? Or in [keras-team/autokeras@49281da](https://github.com/keras-team/autokeras/commit/49281da0a11cd7137818b30a941e88759b6e4741)?

**Comment by bhack** [comment]
I meant as TF Addons Is historically the fusion of Tensor contrib comunity and keras-contrib comunity we have two official upstream promotion paths as you can see [in point 3](https://github.com/tensorflow/addons/blob/master/MIGRATION_TO_CORE.md).
So if we all agree to add new image operation in tfa this mean that the natural promotion path Is tf.image not Keras.
Then there Is the second point that we mentioned in this thread. What about autoaugmentation policies API? Is in TF addon and could be candidate upstream in Keras? Is in Autokeras? Is It in experimental Keras?

**Comment by tanzhenyu** [comment]
> I meant as TF Addons Is historically the fusion of Tensor contrib comunity and keras-contrib comunity we have two official upstream promotion paths as you can see [in point 3](https://github.com/tensorflow/addons/blob/master/MIGRATION_TO_CORE.md).
> So if we all agree to add new image operation in tfa this mean that the natural promotion path Is tf.image not Keras.
> Then there Is the second point that we mentioned in this thread. What about autoaugmentation policies API? Is in TF addon and could be candidate upstream in Keras? Is in Autokeras? Is It in experimental Keras?

I agree with @gabrieldemarmiesse, tfa shouldn't rely on Keras-tuner and include learnable policies. It could just be hard-coded. For autoaugmentation policies API it seems a keras-image repo (which will rely on both tfa.image and Keras-tuner, and AutoKeras can rely on that) would be better fit IMHO.

**Comment by bhack** [comment]
I think that we need to discuss this a little bit cause we need to clarify also what Is the difference contributing in keras.experimental and in TF Addons also to route Gsoc students, contributors etc and to not sparsify implementations on multiple repos also cause currently It Is hard to have a [monthly overview across the Tensorflow/Keras ecosystems](https://github.com/tensorflow/community/issues/29#issuecomment-580894400).

**Comment by tanzhenyu** [comment]
> I think that we need to discuss this a little bit cause we need to clarify also what Is the difference contributing in keras.experimental and in TF Addons also to route Gsoc students, contributors etc and to not sparsify implementations on multiple repos also cause currently It Is hard to have a [monthly overview across the Tensorflow/Keras ecosystems](https://github.com/tensorflow/community/issues/29#issuecomment-580894400).

Yep we definitely need this to be modular and architectural. I will attend the next meeting

**Comment by bhack** [comment]
> Yep we definitely need this to be modular and architectural. I will attend the next meeting

We need to involve somebody from the google-research repo to be sure that Google's internal polices will not create a problem for that team to contribute and feed a policy API under `Keras` org. As you can see in this thread many of the mentioned auto-augmentations and ops are coming from the google-research repos. 
See also the already mentioned https://github.com/tensorflow/community/issues/223

**Comment by tanzhenyu** [comment]
Some follow-up on this: Francois and I discussed and decided that the new [keras_cv](https://github.com/keras-team/keras-cv) will include autoaugment and randaugment.

**Comment by seanpmorgan** [comment]
Closing as this is now going to be implemented elsewhere in the ecosystem. Since we've already merged a few components of these we can deprecate them as they're available.

**Comment by Lewington-pitsos** [comment]
It's been close to a year with no word from on high, keras_cv looks pretty sparse still, has there been any news?
